src/transaction.ts

```diff
@@ -0,0 +1,359 @@
+import {
+  BN,
+  defineProperties,
+  bufferToInt,
+  ecrecover,
+  rlphash,
+  publicToAddress,
+  ecsign,
+  toBuffer,
+  rlp,
+} from 'ethereumjs-util'
+import Common from 'ethereumjs-common'
+import { Buffer } from 'buffer'
+import { BufferLike, PrefixedHexString, TxData, TransactionOptions } from './types'
+
+// secp256k1n/2
+const N_DIV_2 = new BN('7fffffffffffffffffffffffffffffff5d576e7357a4501ddfe92f46681b20a0', 16)
+
+/**
+ * An Ethereum transaction.
+ */
+export default class Transaction {
+  public raw!: Buffer[]
+  public nonce!: Buffer
+  public gasLimit!: Buffer
+  public gasPrice!: Buffer
+  public to!: Buffer
+  public value!: Buffer
+  public data!: Buffer
+  public v!: Buffer
+  public r!: Buffer
+  public s!: Buffer
+
+  private _common: Common
+  private _chainId: number
+  private _senderPubKey?: Buffer
+  protected _from?: Buffer
+
+  /**
+   * Creates a new transaction from an object with its fields' values.
+   *
+   * @param data - A transaction can be initialized with its rlp representation, an array containing
+   * the value of its fields in order, or an object containing them by name. If the latter is used,
+   * a `chainId` can also be provided.
+   *
+   * @param opts - The transaction's options, used to indicate the chain and hardfork the
+   * transactions belongs to.
+   *
+   * @example
+   * ```js
+   * const txData = {
+   *   nonce: '0x00',
+   *   gasPrice: '0x09184e72a000',
+   *   gasLimit: '0x2710',
+   *   to: '0x0000000000000000000000000000000000000000',
+   *   value: '0x00',
+   *   data: '0x7f7465737432000000000000000000000000000000000000000000000000000000600057',
+   *   v: '0x1c',
+   *   r: '0x5e1d3a76fbf824220eafc8c79ad578ad2b67d01b0c2425eb1f1347e8f50882ab',
+   *   s: '0x5bd428537f05f9830e93792f90ea6a3e2d1ee84952dd96edbae9f658f831ab13'
+   * };
+   * const tx = new Transaction(txData);
+   * ```
+   */
+  constructor(
+    data: Buffer | PrefixedHexString | BufferLike[] | TxData = {},
+    opts: TransactionOptions = {},
+  ) {
+    // instantiate Common class instance based on passed options
+    if (opts.common) {
+      if (opts.chain) {
+        throw new Error('Instantiation with both opts.common and opts.chain parameter not allowed!')
+      }
+      this._common = opts.common
+    } else {
+      const chain = opts.chain ? opts.chain : 'mainnet'
```

Comment:
@holgerd77 Is this maybe also a good occasion to drop some deprecated options? namely `opts.chain` and `opts.hardfork` which are now covered by `opts.common`

Reply:
That would simplify the constructor a lot. I don't know how much of an impact this change would have though.

Reply:
These ``Common`` and ``chain``/``hardfork`` options were introduced in parallel along with the the ``Common`` class introduction. Reasoning is to be able to use a shared common class more in a programmatic/dynamic context as well as keep a simple option for users who just want to instantiate e.g. a ``chain:goerli`` tx without having to deal with the extra Common dependency.

Not sure if this is ideal but this has been introduced along all libraries using the Common library as a shared pattern so for now I would like to stick with that if there is no strong reasoning against.

Reply:
Oh, sorry I thought they were for backwards compatibility. No I don't feel very strongly about this.